src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -166,40 +166,48 @@ private bool TryRewriteSingleElementSpreadToList(BoundCollectionExpression node,
                 return false;
             }
 
+            if (!ShouldUseAddRangeOrToListMethod(singleSpread.Expression.Type, singleSpread.EnumeratorInfoOpt?.GetEnumeratorInfo.Method, listElementType))
+            {
+                return false;
+            }
+
+            if (TryGetWellKnownTypeMember(node.Syntax, WellKnownMember.System_Linq_Enumerable__ToList, out MethodSymbol? toListGeneric, isOptional: true))
+            {
+                var toListOfElementType = toListGeneric.Construct([listElementType]);
+                result = _factory.Call(receiver: null, toListOfElementType, singleSpread.Expression);
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool ShouldUseAddRangeOrToListMethod(TypeSymbol? spreadType, MethodSymbol? getEnumeratorMethod, TypeWithAnnotations elementType)
+        {
+            Debug.Assert(spreadType is not null);
```

Comment:
We can leave this assertion here, but, I would prefer if the parameter had non-nullable type, and the call sites put `!` on the argument. It's better for the function signature to indicate whether the parameter is allowed to be null.